### PR TITLE
Bluetooth: Mesh: Refactor Config Client and Health Client API to allow async use

### PR DESCRIPTION
**Pull request by michal-narajowski**
- Don't ignore unsolicited status messages to allow async use in Health Client and Config Client
- Split Unacknowledged API in Health Client
- Add more callbacks in Health Client to notify applications of status messages

### CODE REVIEW

**Comment by LingaoM** [comment]
This PR gives me an idea:
At present, All mesh messages requiring ack are processed synchronously through block k_sem_take.
Can all be changed to asynchronous mode and called through callback.
Is this more flexible for application layer development?

**Comment by LingaoM** [comment]
> Can all be changed to asynchronous mode and called through callback.

After careful consideration:

This method needs to define all mesh config client status callbacks & health client status callbacks (but it is still const. For some messages that are not planned to be used, the application layer can of course be assigned NULL, although it still wastes 4 bytes of ROM space),

Another problem is that we have defined all cfg_cli status processing functions at present. Even if we do not intend to use them, it still exists. Therefore, it can be changed to a general field analysis function, which can be called by the upper layer at the application layer through callback. We do not use some status processing logic, so it will be optimized away.

For the application layer, it is completely possible to choose asynchronous or synchronous processing, which has more freedom than the current point of view.

**Comment by michal-narajowski** [comment]
@LingaoM I'm not sure I fully understand your proposal, so let me rephrase it:

Instead of having Status processing functions in cfg_cli.c we notify the application that a status message is received and pass raw data and a callback that will process the status message and the application will decide whether it wants to process it?

**Comment by LingaoM** [comment]
> @LingaoM I'm not sure I fully understand your proposal, so let me rephrase it:
> 
> 
> 
> Instead of having Status processing functions in cfg_cli.c we notify the application that a status message is received and pass raw data and a callback that will process the status message and the application will decide whether it wants to process it?

Sure.

**Review by trond-snekvik — CHANGES_REQUESTED** [review]
Looks good to me, apart from this minor comment. Well done avoiding any changes to the existing API👍 

> At present, All mesh messages requiring ack are processed synchronously through block k_sem_take.
Can all be changed to asynchronous mode and called through callback.
Is this more flexible for application layer development?

We've implemented something very close to this pattern in [Nordic's nRF Connect SDK](https://github.com/nrfconnect/sdk-nrf/tree/main/include/bluetooth/mesh), but we have one callback per message type, and always call the app callbacks, as well as signaling any blocking semaphores, pretty much the same way as implemented in the health client in this PR.

If we want to add support for asynchronous modes for the config client, I believe this should be the general pattern, so we can receive status messages in both ways at the same time. For models that also receive asynchronous status publications, it's the only pattern that makes sense, I think.

**Review by trond-snekvik — APPROVED** [review]

**Review by jhedberg — APPROVED** [review]

**Review by LingaoM — APPROVED** [review]

**Comment by WilliamGFish** [comment]
unable to get provisioning sample to function, fails to retrieve node composition data `_(Failed to get Composition data (err 0, status: 241)_`

Was working prior to this change.

Billy..
continuing to investigate

File: subsys/bluetooth/mesh/cfg_cli.c [hunk 1]
```diff
@@ -43,26 +43,26 @@ static int comp_data_status(struct bt_mesh_model *model,
...
 
-	if (!bt_mesh_msg_ack_ctx_match(&cli->ack_ctx, OP_DEV_COMP_DATA_STATUS, ctx->addr,
-				       (void **)&param)) {
-		return -ENOENT;
-	}
-
-	if (param->page) {
-		*(param->page) = net_buf_simple_pull_u8(buf);
-	}
+	status = net_buf_simple_pull_u8(buf);
```
Comment (WilliamGFish) [inline subsys/bluetooth/mesh/cfg_cli.c:53]: By pulling the status/page (u8) here and then pull an additional u8 if there is a page parameter the Composition Status will always be malformed.

Could simply rework to 
```
		if (param->page) {
			*(param->page) = net_buf_simple_pull_u8(buf);

		}else{
			status = net_buf_simple_pull_u8(buf);
		}
```

**Review by WilliamGFish — COMMENTED** [review]
This change stops mesh provisioning, returns an invalid 'status' in message.

File: subsys/bluetooth/mesh/cfg_cli.c (hunk 1 above)
Comment (michal-narajowski) [reply subsys/bluetooth/mesh/cfg_cli.c:53]: Good point, I think I messed it up when I was resolving conflicts. I will post a patch asap.

Comment (WilliamGFish) [reply subsys/bluetooth/mesh/cfg_cli.c:53]: @michal-narajowski 
Don't worry that's why people like me run changes through real world tests just to double check, community effort.

Billy..
